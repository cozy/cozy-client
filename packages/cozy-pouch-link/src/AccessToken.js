export default class AccessToken {
  constructor(opts) {
    this.tokenType = opts.tokenType || opts.token_type
    this.accessToken = opts.accessToken || opts.access_token
    this.refreshToken = opts.refreshToken || opts.refresh_token
    this.scope = opts.scope
  }

  toAuthHeader() {
    return 'Bearer ' + this.accessToken
  }

  toBasicAuth() {
    return `user:${this.accessToken}@`
  }

  static fromJSON(data) {
    return new AccessToken({
      tokenType: data.tokenType,
      accessToken: data.accessToken,
      refreshToken: data.refreshToken,
      scope: data.scope
    })
  }

  toJSON() {
    return {
      tokenType: this.tokenType,
      accessToken: this.accessToken,
      refreshToken: this.refreshToken,
      scope: this.scope
    }
  }
}
