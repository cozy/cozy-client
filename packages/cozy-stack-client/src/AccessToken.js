export default class AccessToken {
  constructor(data) {
    if (typeof data === 'string') data = JSON.parse(data)

    this.tokenType = data.token_type || data.tokenType
    this.accessToken = data.access_token || data.accessToken
    this.refreshToken = data.refresh_token || data.refreshToken
    this.scope = data.scope
  }

  toAuthHeader() {
    return 'Bearer ' + this.accessToken
  }

  toBasicAuth() {
    return `user:${this.accessToken}@`
  }

  toJSON() {
    return {
      tokenType: this.tokenType,
      accessToken: this.accessToken,
      refreshToken: this.refreshToken,
      scope: this.scope
    }
  }

  toString() {
    return JSON.stringify(this.toJSON())
  }
}
