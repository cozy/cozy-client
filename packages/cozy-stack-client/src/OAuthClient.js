import CozyStackClient from './CozyStackClient'
import AccessToken from './AccessToken'

const defaultoauthOptions = {
  clientID: '',
  clientName: '',
  clientKind: '',
  clientSecret: '',
  clientURI: '',
  registrationAccessToken: '',
  redirectURI: '',
  softwareID: '',
  softwareVersion: '',
  logoURI: '',
  policyURI: '',
  notificationPlatform: '',
  notificationDeviceToken: ''
}

export default class OAuthClient extends CozyStackClient {
  constructor({ oauth, scope = [], ...options }) {
    super(options)
    this.setOAuthOptions({ ...defaultoauthOptions, ...oauth })
    this.scope = scope
  }

  /**
   * Checks if the client has his registration information from the server
   * @returns {boolean}
   * @private
   */
  isRegistered() {
    return this.oauthOptions.clientID !== ''
  }

  /**
   * Converts a camel-cased data set to snake case, suitable for sending to the OAuth server
   * @param   {object} data Initial data
   * @returns {object} Formatted data
   * @private
   */
  snakeCaseOAuthData(data) {
    const mappedFields = {
      softwareID: 'software_id',
      softwareVersion: 'software_version',
      clientID: 'client_id',
      clientName: 'client_name',
      clientKind: 'client_kind',
      clientURI: 'client_uri',
      logoURI: 'logo_uri',
      policyURI: 'policy_uri',
      notificationPlatform: 'notification_platform',
      notificationDeviceToken: 'notification_device_token',
      redirectURI: 'redirect_uris'
    }

    const result = {}

    Object.keys(data).forEach(fieldName => {
      const key = mappedFields[fieldName] || fieldName
      const value = data[fieldName]
      result[key] = value
    })

    // special case: turn redirect_uris into an array
    if (
      result['redirect_uris'] &&
      result['redirect_uris'] instanceof Array === false
    )
      result['redirect_uris'] = [result['redirect_uris']]

    return result
  }

  /**
   * Converts a snake-cased data set to camel case, suitable for internal use
   * @param   {object} data Initial data
   * @returns {object} Formatted data
   * @private
   */
  camelCaseOAuthData(data) {
    const mappedFields = {
      client_id: 'clientID',
      client_name: 'clientName',
      client_secret: 'clientSecret',
      registration_access_token: 'registrationAccessToken',
      software_id: 'softwareID'
    }

    const result = {}

    Object.keys(data).forEach(fieldName => {
      const key = mappedFields[fieldName] || fieldName
      const value = data[fieldName]
      result[key] = value
    })

    return result
  }

  /**
   * Registers the currenly configured client with the OAuth server.
   * @throws {Error} When the client is already registered
   * @returns {promise} A promise that resolves with a complete list of client information, including client ID and client secret.
   */
  async register() {
    if (this.isRegistered()) throw new Error('Client already registered')

    const data = await this.fetch(
      'POST',
      '/auth/register',
      this.snakeCaseOAuthData({
        redirectURI: this.oauthOptions.redirectURI,
        clientName: this.oauthOptions.clientName,
        softwareID: this.oauthOptions.softwareID,
        clientKind: this.oauthOptions.clientKind,
        clientURI: this.oauthOptions.clientURI,
        logoURI: this.oauthOptions.logoURI,
        policyURI: this.oauthOptions.policyURI,
        softwareVersion: this.oauthOptions.softwareVersion,
        notificationPlatform: this.oauthOptions.notificationPlatform,
        notificationDeviceToken: this.oauthOptions.notificationDeviceToken
      })
    )

    this.setOAuthOptions({
      ...this.oauthOptions,
      client_id: data.client_id,
      client_name: data.client_name,
      client_secret: data.client_secret,
      registration_access_token: data.registration_access_token,
      software_id: data.software_id
    })

    return this.oauthOptions
  }

  /**
   * Unregisters the currenly configured client with the OAuth server.
   * @throws {NotRegisteredException} When the client doesn't have it's registration information
   * @returns {promise}
   */
  async unregister() {
    if (!this.isRegistered()) throw new NotRegisteredException()

    return this.fetch(
      'DELETE',
      `/auth/register/${this.oauthOptions.clientID}`,
      null,
      {
        credentials: this.registrationAccessTokenToAuthHeader()
      }
    )
  }

  /**
   * Fetches the complete set of client information from the server after it has been registered.
   * @throws {NotRegisteredException} When the client doesn't have it's registration information
   * @returns {promise}
   */
  async fetchInformation() {
    if (!this.isRegistered()) throw new NotRegisteredException()

    return this.fetch(
      'GET',
      `/auth/register/${this.oauthOptions.clientID}`,
      null,
      {
        credentials: this.registrationAccessTokenToAuthHeader()
      }
    )
  }

  /**
   * Updates the client own information. This method will update both the local information and the remote information on the OAuth server.
   * @throws {NotRegisteredException} When the client doesn't have it's registration information
   * @param   {object} information Set of information to update. Note that some fields such as `clientID` can't be updated.
   * @param   {boolean} resetSecret = false Optionnal, whether to reset the client secret or not
   * @returns {promise} A promise that resolves to a complete, updated list of client information
   */
  async updateInformation(information, resetSecret = false) {
    if (!this.isRegistered()) throw new NotRegisteredException()

    const mandatoryFields = {
      clientID: this.oauthOptions.clientID,
      clientName: this.oauthOptions.clientName,
      redirectURI: this.oauthOptions.redirectURI,
      softwareID: this.oauthOptions.softwareID
    }
    const data = this.snakeCaseOAuthData({
      ...mandatoryFields,
      ...information
    })

    if (resetSecret) data['client_secret'] = this.oauthOptions.clientSecret

    const result = await this.fetch(
      'PUT',
      `/auth/register/${this.oauthOptions.clientID}`,
      data,
      {
        credentials: this.registrationAccessTokenToAuthHeader()
      }
    )

    this.setOAuthOptions(result)

    return this.oauthOptions
  }

  /**
   * Generates a random state code to be used during the OAuth process
   * @returns {string}
   */
  generateStateCode() {
    const STATE_SIZE = 16
    const hasCrypto =
      typeof window !== 'undefined' &&
      typeof window.crypto !== 'undefined' &&
      typeof window.crypto.getRandomValues === 'function'

    let buffer
    if (hasCrypto) {
      buffer = new Uint8Array(STATE_SIZE)
      window.crypto.getRandomValues(buffer)
    } else {
      buffer = new Array(STATE_SIZE)
      for (let i = 0; i < buffer.length; i++) {
        buffer[i] = Math.floor(Math.random() * 255)
      }
    }

    return btoa(String.fromCharCode.apply(null, buffer))
      .replace(/=+$/, '')
      .replace(/\//g, '_')
      .replace(/\+/g, '-')
  }

  /**
   * Generates the URL that the user should be sent to in order to accept the app's permissions.
   * @throws {NotRegisteredException} When the client doesn't have it's registration information
   * @param   {string} stateCode   A random code to be included in the URl for security. Can be generated with `client.generateStateCode()`
   * @param   {Array} scopes = [] An array of permission scopes for the token.
   * @returns {string} The URL
   */
  getAuthCodeURL(stateCode, scopes = this.scope) {
    if (!this.isRegistered()) throw new NotRegisteredException()

    const query = new URLSearchParams({
      client_id: this.oauthOptions.clientID,
      redirect_uri: this.oauthOptions.redirectURI,
      state: stateCode,
      response_type: 'code',
      scope: scopes.join(' ')
    })

    return `${this.uri}/auth/authorize?${query.toString()}`
  }

  /**
   * Retrieves the access code contained in the URL to which the user is redirected after accepting the app's permissions (the `redirectURI`).
   * @throws {Error} The URL should contain the same state code as the one generated with `client.getAuthCodeURL()`. If not, it will throw an error
   * @param   {string} pageURL The redirected page URL, containing the state code and the access code
   * @param   {string} stateCode The state code that was contained in the original URL the user was sent to (see `client.getAuthCodeURL()`)
   * @returns {string} The access code
   */
  getAccessCodeFromURL(pageURL, stateCode) {
    if (!stateCode) throw new Error('Missing state code')

    const params = new URL(pageURL).searchParams
    const urlStateCode = params.get('state')
    const urlAccessCode = params.get('access_code')

    if (stateCode !== urlStateCode)
      throw new Error('Given state does not match url query state')

    return urlAccessCode
  }

  /**
   * Exchanges an access code for an access token. This function does **not** update the client's token.
   * @throws {NotRegisteredException} When the client doesn't have it's registration information
   * @param   {string} accessCode The access code contained in the redirection URL â€” sett `client.getAccessCodeFromURL()`
   * @returns {Promise} A promise that resolves with an AccessToken object.
   */
  async fetchAccessToken(accessCode) {
    if (!this.isRegistered()) throw new NotRegisteredException()

    const data = new URLSearchParams({
      grant_type: 'authorization_code',
      code: accessCode,
      client_id: this.oauthOptions.clientID,
      client_secret: this.oauthOptions.clientSecret
    })

    const result = await this.fetch('POST', '/auth/access_token', data, {
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
    })

    return new AccessToken(result)
  }

  /**
   * Retrieves a new access token by refreshing the currently used token.
   * @throws {NotRegisteredException} When the client doesn't have it's registration information
   * @throws {Error} The client should already have an access token to use this function
   * @returns {Promise} A promise that resolves with a new AccessToen object
   */
  async refreshToken() {
    if (!this.isRegistered()) throw new NotRegisteredException()
    if (!this.token) throw new Error('No token to refresh')

    const data = new URLSearchParams({
      grant_type: 'refresh_token',
      refresh_token: this.token.refreshToken,
      client_id: this.oauthOptions.clientID,
      client_secret: this.oauthOptions.clientSecret
    })

    const result = await this.fetch('POST', '/auth/access_token', data, {
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
    })

    const newToken = new AccessToken({
      refresh_token: this.token.refreshToken,
      ...result
    })

    if (
      this.oauthOptions.onTokenRefresh &&
      typeof this.oauthOptions.onTokenRefresh === 'function'
    ) {
      this.oauthOptions.onTokenRefresh(newToken)
    }

    return newToken
  }

  /**
   * Updates the client's stored token
   * @param {string} token = null The new token to use â€” can be a string, a json object or an AccessToken instance.
   */
  setCredentials(token) {
    if (token) {
      this.token = token instanceof AccessToken ? token : new AccessToken(token)
    } else {
      this.token = null
    }
  }
  
  /**
   * Updates the OAuth informations
   * @param {object} options Map of OAuth options
   */
  setOAuthOptions(options) {
    this.oauthOptions = this.camelCaseOAuthData(options)
  }
  
  resetClientId() {
    this.oauthOptions.clientID = ''
  }

  async fetch(method, path, body, options) {
    try {
      return super.fetch(method, path, body, options)
    } catch (e) {
      if (/Expired token/.test(e.message)) {
        const token = await this.refreshToken()
        this.setCredentials(token)
        return await super.fetch(method, path, body, options)
      } else {
        throw e
      }
    }
  }

  /**
   * Turns the client's registration access token into a header suitable for HTTP requests. Used in some queries to manipulate the client on the server side.
   * @returns {string}
   * @private
   */
  registrationAccessTokenToAuthHeader() {
    return 'Bearer ' + this.oauthOptions.registrationAccessToken
  }
}

class NotRegisteredException extends Error {
  constructor(message = 'Client not registered or missing OAuth information') {
    super(message)
    this.message = message
    this.name = 'NotRegisteredException'
  }
}
