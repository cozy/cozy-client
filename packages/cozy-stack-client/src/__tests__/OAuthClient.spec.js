import OAuthClient from '../OAuthClient'
import AccessToken from '../AccessToken'

const CLIENT_INIT_OPTIONS = {
  uri: 'http://cozy.tools:8080',
  oauth: {
    clientName: 'TestClientName',
    softwareID: 'TestSofwareID',
    redirectURI: 'http://localhost'
  }
}

const REGISTERED_CLIENT_INIT_OPTIONS = {
  uri: CLIENT_INIT_OPTIONS.uri,
  oauth: {
    ...CLIENT_INIT_OPTIONS.oauth,
    clientID: '1',
    clientsecret: '1',
    registrationAccessToken: '1234'
  }
}

describe('OAuthClient', () => {
  beforeAll(() => {
    global.fetch = require('jest-fetch-mock')
  })

  beforeEach(() => {
    fetch.resetMocks()
    fetch.mockResponse(JSON.stringify({ foo: 'bar' }))
  })

  let client

  describe('without registration', () => {
    beforeEach(() => {
      client = new OAuthClient(CLIENT_INIT_OPTIONS)
    })

    it('should register a client', () => {
      client.register()
      expect(fetch.mock.calls[0]).toMatchSnapshot()
    })

    it('should throw on other server interactions', () => {
      expect(client.unregister()).rejects.toThrowErrorMatchingSnapshot()
      expect(client.fetchInformation()).rejects.toThrowErrorMatchingSnapshot()
      expect(client.updateInformation()).rejects.toThrowErrorMatchingSnapshot()
      expect(client.fetchAccessToken()).rejects.toThrowErrorMatchingSnapshot()
      expect(client.refreshToken()).rejects.toThrowErrorMatchingSnapshot()
    })

    it('should not generate a auth code URL', () => {
      expect(() => {
        client.getAuthCodeURL('statecode')
      }).toThrowErrorMatchingSnapshot()
    })
  })

  describe('with registration', () => {
    beforeEach(() => {
      client = new OAuthClient(REGISTERED_CLIENT_INIT_OPTIONS)
      client.setCredentials({
        tokenType: 'type',
        accessToken: 'abcd',
        refreshToken: 'refresh-789',
        scope: 'io.cozy.todos'
      })
    })

    it('should throw when trying to register again', () => {
      expect(client.register()).rejects.toThrowErrorMatchingSnapshot()
    })

    it('should unregister a client', () => {
      client.unregister()
      expect(fetch.mock.calls[0]).toMatchSnapshot()
    })

    it('should fetch client informations', () => {
      client.fetchInformation()
      expect(fetch.mock.calls[0]).toMatchSnapshot()
    })

    it('should update client informations', () => {
      client.updateInformation({
        policy_url: 'http://example.com'
      })
      expect(fetch.mock.calls[0]).toMatchSnapshot()
    })

    it('should generate a random state code', () => {
      expect(client.generateStateCode()).toBeDefined()
    })

    it('should generate the auth code URL', () => {
      expect(
        client.getAuthCodeURL('randomstatetoken', ['io.cozy.todos'])
      ).toEqual(
        `${REGISTERED_CLIENT_INIT_OPTIONS.uri}/auth/authorize?client_id=${
          REGISTERED_CLIENT_INIT_OPTIONS.oauth.clientID
        }&redirect_uri=${encodeURIComponent(
          REGISTERED_CLIENT_INIT_OPTIONS.oauth.redirectURI
        )}&state=randomstatetoken&response_type=code&scope=io.cozy.todos`
      )
    })

    it('should get the access code from an URL', () => {
      const stateCode = 'myrandomcode'
      const accessCode = 'myaccesscode'
      const url = `http://example.com?state=${stateCode}&access_code=${accessCode}`
      expect(client.getAccessCodeFromURL(url, stateCode)).toEqual(accessCode)
    })

    it('should throw when no access code is provided', () => {
      expect(() => {
        client.getAccessCodeFromURL('http://example.com')
      }).toThrowErrorMatchingSnapshot()
    })

    it('should throw when the provided access code is different from the URL', () => {
      const stateCode = 'myrandomcode'
      const accessCode = 'myaccesscode'
      const url = `http://example.com?state=${stateCode}&access_code=${accessCode}`

      expect(() => {
        client.getAccessCodeFromURL(url, 'incorrect')
      }).toThrowErrorMatchingSnapshot()
    })

    it('should fetch an access token', () => {
      client.fetchAccessToken('myaccesstoken')
      expect(fetch.mock.calls[0]).toMatchSnapshot()
    })

    it('should refresh the access token', () => {
      client.refreshToken()
      expect(fetch.mock.calls[0]).toMatchSnapshot()
    })

    it('should throw when refreshing the access token without token', () => {
      client = new OAuthClient(REGISTERED_CLIENT_INIT_OPTIONS)
      expect(client.refreshToken()).rejects.toThrowErrorMatchingSnapshot()
    })

    it('should update the internal credentials', () => {
      const camelCredentials = {
        tokenType: 'Camel',
        accessToken: 'Cased',
        refreshToken: 'refresh-me',
        scope: 'io.cozy.todos'
      }
      const snakeCredentials = {
        token_type: 'Snake',
        access_token: 'DoubleCased',
        refresh_token: 'refresh-you',
        scope: 'io.cozy.todos'
      }
      client.setCredentials(camelCredentials)
      expect(client.token).toBeInstanceOf(AccessToken)
      expect(client.token.tokenType).toEqual(camelCredentials.tokenType)
      expect(client.token.accessToken).toEqual(camelCredentials.accessToken)
      expect(client.token.refreshToken).toEqual(camelCredentials.refreshToken)
      expect(client.token.scope).toEqual(camelCredentials.scope)

      client.setCredentials(JSON.stringify(snakeCredentials))
      expect(client.token).toBeInstanceOf(AccessToken)
      expect(client.token.tokenType).toEqual(snakeCredentials.token_type)
      expect(client.token.accessToken).toEqual(snakeCredentials.access_token)
      expect(client.token.refreshToken).toEqual(snakeCredentials.refresh_token)
      expect(client.token.scope).toEqual(snakeCredentials.scope)
    })

    it('should call getCredentials for usual requests', () => {
      const spy = jest.spyOn(client, 'getCredentials')
      client.fetch('GET', 'http://example.com')
      expect(spy).toHaveBeenCalled()
    })
  })
})
