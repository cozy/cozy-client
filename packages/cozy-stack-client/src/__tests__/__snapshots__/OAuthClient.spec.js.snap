// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OAuthClient with registration should fetch an access token 1`] = `
Array [
  "http://cozy.tools:8080/auth/access_token",
  Object {
    "body": URLSearchParams {
      Symbol(impl): URLSearchParamsImpl {
        "_list": Array [
          Array [
            "grant_type",
            "authorization_code",
          ],
          Array [
            "code",
            "myaccesstoken",
          ],
          Array [
            "client_id",
            "1",
          ],
          Array [
            "client_secret",
            "",
          ],
        ],
        "_url": null,
        Symbol(wrapper): [Circular],
      },
    },
    "credentials": "include",
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer abcd",
      "Content-Type": "application/x-www-form-urlencoded",
    },
    "method": "POST",
  },
]
`;

exports[`OAuthClient with registration should fetch client informations 1`] = `
Array [
  "http://cozy.tools:8080/auth/register/1",
  Object {
    "credentials": "include",
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer 1234",
    },
    "method": "GET",
  },
]
`;

exports[`OAuthClient with registration should refresh the access token 1`] = `
Array [
  "http://cozy.tools:8080/auth/access_token",
  Object {
    "body": URLSearchParams {
      Symbol(impl): URLSearchParamsImpl {
        "_list": Array [
          Array [
            "grant_type",
            "refresh_token",
          ],
          Array [
            "refresh_token",
            "refresh-789",
          ],
          Array [
            "client_id",
            "1",
          ],
          Array [
            "client_secret",
            "",
          ],
        ],
        "_url": null,
        Symbol(wrapper): [Circular],
      },
    },
    "credentials": "include",
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer abcd",
      "Content-Type": "application/x-www-form-urlencoded",
    },
    "method": "POST",
  },
]
`;

exports[`OAuthClient with registration should throw when no access code is provided 1`] = `"Missing state code"`;

exports[`OAuthClient with registration should throw when refreshing the access token without token 1`] = `"No token to refresh"`;

exports[`OAuthClient with registration should throw when the provided access code is different from the URL 1`] = `"Given state does not match url query state"`;

exports[`OAuthClient with registration should throw when trying to register again 1`] = `"Client already registered"`;

exports[`OAuthClient with registration should unregister a client 1`] = `
Array [
  "http://cozy.tools:8080/auth/register/1",
  Object {
    "body": "null",
    "credentials": "include",
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer 1234",
      "Content-Type": "application/json",
    },
    "method": "DELETE",
  },
]
`;

exports[`OAuthClient with registration should update client informations 1`] = `
Array [
  "http://cozy.tools:8080/auth/register/1",
  Object {
    "body": "{\\"client_id\\":\\"1\\",\\"client_name\\":\\"TestClientName\\",\\"redirect_uris\\":[\\"http://localhost\\"],\\"software_id\\":\\"TestSofwareID\\",\\"policy_url\\":\\"http://example.com\\"}",
    "credentials": "include",
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer 1234",
      "Content-Type": "application/json",
    },
    "method": "PUT",
  },
]
`;

exports[`OAuthClient without registration should not generate a auth code URL 1`] = `"Client not registered or missing OAuth information"`;

exports[`OAuthClient without registration should register a client 1`] = `
Array [
  "http://cozy.tools:8080/auth/register",
  Object {
    "body": "{\\"redirect_uris\\":[\\"http://localhost\\"],\\"client_name\\":\\"TestClientName\\",\\"software_id\\":\\"TestSofwareID\\",\\"client_kind\\":\\"\\",\\"client_uri\\":\\"\\",\\"logo_uri\\":\\"\\",\\"policy_uri\\":\\"\\",\\"software_version\\":\\"\\",\\"notification_platform\\":\\"\\",\\"notification_device_token\\":\\"\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`OAuthClient without registration should throw on other server interactions 1`] = `"Client not registered or missing OAuth information"`;

exports[`OAuthClient without registration should throw on other server interactions 2`] = `"Client not registered or missing OAuth information"`;

exports[`OAuthClient without registration should throw on other server interactions 3`] = `"Client not registered or missing OAuth information"`;

exports[`OAuthClient without registration should throw on other server interactions 4`] = `"Client not registered or missing OAuth information"`;

exports[`OAuthClient without registration should throw on other server interactions 5`] = `"Client not registered or missing OAuth information"`;
